File: src/game/combat.py
"""
攻击力/防御力计算，破坏与连通分量结算，直接攻击，红/蓝/绿效果。
"""
from __future__ import annotations
from typing import Set, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from src.game.state import GameState

from src.grid.cell import Cell, ATOM_BLACK, ATOM_RED, ATOM_BLUE, ATOM_GREEN
from src.grid.triangle import vertical_distance_units, horizontal_distance_units, GridPoint


def attack_power(cell: Cell) -> float:
    """进攻格攻击力：最高与最低黑原子的竖向距离（单位：三角形高 √3/2）。"""
    blacks = cell.black_points()
    return vertical_distance_units(blacks)


def defense_power(cell: Cell) -> float:
    """防守格防御力：最左与最右黑原子的横向距离（单位：边长 1）。"""
    blacks = cell.black_points()
    return horizontal_distance_units(blacks)


def attack_beats_defense(attacker: Cell, defender: Cell) -> bool:
    return attack_power(attacker) > defense_power(defender)


def extra_destroys(attacker_cell: Cell, defender_cell: Cell) -> int:
    """
    进攻格红原子数 n、防守格蓝原子数 m 时，在破坏 1 个黑原子后还可多破坏的原子数。
    规则：可破坏数 = max(0, 1 + n - m)，已破坏 1 黑，故额外 = max(0, n - m)。
    """
    n = attacker_cell.count_by_color().get(ATOM_RED, 0)
    m = defender_cell.count_by_color().get(ATOM_BLUE, 0)
    return max(0, n - m)


def destroy_atom(cell: Cell, r: int, c: int) -> Optional[str]:
    """移除格点上的原子，返回被移除的颜色。"""
    return cell.remove(r, c)


def destroy_one_black_and_get_components(
    attacker_cell: Cell,
    defender_cell: Cell,
    black_to_destroy: GridPoint,
) -> tuple[int, List[Set[GridPoint]]]:
    """
    只破坏防守格中指定的一个黑原子，不自动处理连通分量。
    返回 (对玩家造成的伤害, 破坏后的连通分量列表)。
    伤害为 1（破坏 1 个黑原子）。
    """
    if defender_cell.get(black_to_destroy[0], black_to_destroy[1]) != ATOM_BLACK:
        return 0, []
    defender_cell.remove(black_to_destroy[0], black_to_destroy[1])
    components = defender_cell.connected_components()
    return 1, components


def remove_components_except(cell: Cell, to_keep: Set[GridPoint]) -> None:
    """保留 to_keep 所在连通分量，移除其余分量上的所有原子。"""
    components = cell.connected_components()
    for comp in components:
        if to_keep & comp:
            continue
        for (r, c) in comp:
            cell.remove(r, c)


def remove_component(cell: Cell, component: Set[GridPoint]) -> None:
    """移除该连通分量上的所有原子（红效果选到无黑子集时整块破坏）。"""
    for (r, c) in component:
        cell.remove(r, c)


def remove_black_atoms_except(cell: Cell, to_keep: Set[GridPoint]) -> None:
    """仅移除黑原子：保留 to_keep 中的黑原子，移除其余格点上的黑原子。"""
    blacks = cell.black_points()
    for pt in blacks:
        if pt not in to_keep:
            cell.remove(pt[0], pt[1])


def clear_cell_if_no_black(cell: Cell) -> None:
    """若格子内无黑原子则清空整格（攻击/红效果结算后调用）。"""
    if not cell.has_black() and not cell.is_empty():
        for p in list(cell.all_atoms().keys()):
            cell.remove(p[0], p[1])


def remove_components_without_black_and_return_rest(cell: Cell) -> List[Set[GridPoint]]:
    """
    移除所有不包含黑原子的连通分量；返回剩余分量（每个都含黑）。
    若剩余 0 或 1 个则无需被攻击方选择；2+ 个时需选择保留哪一个。
    """
    if cell.is_empty():
        return []
    components = cell.connected_components()
    blacks = cell.black_points()
    remaining: List[Set[GridPoint]] = []
    for comp in components:
        if comp & blacks:
            remaining.append(comp)
        else:
            for (r, c) in comp:
                cell.remove(r, c)
    return remaining


def resolve_direct_attack(cell: Cell) -> int:
    """直接攻击玩家时造成的伤害 = 进攻格攻击力（整数）。"""
    return int(attack_power(cell))


def apply_effect_red(
    state: GameState,
    attacker: int,
    cell_index: int,
    r: int,
    c: int,
    atoms_to_destroy: List[tuple[int, int, GridPoint]],
) -> bool:
    """
    发动红效果：移除己方 (cell_index, r, c) 上的红原子；
    对方必须破坏若干黑原子，atoms_to_destroy 为 [(defender, cell_i, (r,c)), ...]，
    数量为 min(y, 对方当前黑原子数)，且每个目标必须为黑原子。
    若某格破坏后无黑，该格全部原子破坏。
    """
    cells = state.cells[attacker]
    if cell_index < 0 or cell_index >= len(cells):
        return False
    cell = cells[cell_index]
    if cell.get(r, c) != ATOM_RED:
        return False
    y = cell.count_black_neighbors(r, c)
    if len(atoms_to_destroy) > y:
        return False
    defender = state.opponent(attacker)
    def_cells = state.cells[defender]
    for (pi, ci, pt) in atoms_to_destroy:
        if pi != defender or ci < 0 or ci >= len(def_cells):
            return False
        if def_cells[ci].get(pt[0], pt[1]) != ATOM_BLACK:
            return False
    cell.remove(r, c)
    for (pi, ci, pt) in atoms_to_destroy:
        def_cells[ci].remove(pt[0], pt[1])
    # 结算后：无黑则整格清空；多分量时先清掉不含黑的子集，再让被攻击方选择保留哪一子集（仅当仍剩 2+ 个子集时）
    for cell_def in def_cells:
        clear_cell_if_no_black(cell_def)
    for cell_def in def_cells:
        remove_components_without_black_and_return_rest(cell_def)
    return True


def apply_effect_blue(state: GameState, player: int, cell_index: int, r: int, c: int) -> bool:
    """发动蓝效果：移除该蓝原子，与其相邻的黑原子在下一回合内不可被破坏。"""
    cells = state.cells[player]
    if cell_index < 0 or cell_index >= len(cells):
        return False
    cell = cells[cell_index]
    if cell.get(r, c) != ATOM_BLUE:
        return False
    protected = cell.black_neighbors_of(r, c)
    cell.remove(r, c)
    for pt in protected:
        state.blue_protected_points[player].add((cell_index, pt))
    state.blue_protection_until_turn[player] = state.turn_number + 1
    return True


def apply_effect_green(state: GameState, player: int, cell_index: int, r: int, c: int) -> bool:
    """发动绿效果（点击）：该绿原子就地变为一个黑原子，不消耗池中黑原子。"""
    cells = state.cells[player]
    if cell_index < 0 or cell_index >= len(cells):
        return False
    cell = cells[cell_index]
    if cell.get(r, c) != ATOM_GREEN:
        return False
    cell.remove(r, c)
    cell.place(r, c, ATOM_BLACK)
    return True


def apply_green_end_of_turn(state: GameState, player: int) -> int:
    """绿持续效果：回合结束时该玩家获得「己方所有绿原子邻接黑原子数」之和的黑原子。返回本次获得的数目。"""
    total = 0
    for cell in state.cells[player]:
        for (r, c), color in list(cell.all_atoms().items()):
            if color == ATOM_GREEN:
                total += cell.count_black_neighbors(r, c)
    if total > 0:
        state.pools[player][ATOM_BLACK] = state.pools[player].get(ATOM_BLACK, 0) + total
    return total

File: web/src/game/combat.js
/**
 * Attack/defense, destroy, connectivity, effects.
 */
import { verticalDistanceUnits, horizontalDistanceUnits } from './grid.js'
import { ATOM_BLACK, ATOM_RED, ATOM_BLUE, ATOM_GREEN, ATOM_YELLOW, ATOM_PURPLE } from './config.js'

export function attackPower(cell) {
  const blacks = cell.blackPoints()
  return verticalDistanceUnits(blacks)
}

export function defensePower(cell) {
  const blacks = cell.blackPoints()
  return horizontalDistanceUnits(blacks)
}

export function attackBeatsDefense(attacker, defender) {
  return attackPower(attacker) > defensePower(defender)
}

/**
 * 蓝原子持续性：遭到进攻时，己方格蓝原子连接黑原子数目为 x，则降低可以破坏的原子数目 x（最低为 0）。
 * 红方贡献 = 各红原子有效黑邻数之和（与紫相邻时用两跳黑邻）；蓝方 x = 各蓝原子有效黑邻数之和。
 */
export function extraDestroys(attackerCell, defenderCell) {
  let redSum = 0
  for (const [[r, c], color] of attackerCell.allAtoms()) {
    if (color === ATOM_RED) redSum += attackerCell.effectiveBlackNeighborCount(r, c)
  }
  let blueSum = 0
  for (const [[r, c], color] of defenderCell.allAtoms()) {
    if (color === ATOM_BLUE) blueSum += defenderCell.effectiveBlackNeighborCount(r, c)
  }
  return Math.max(0, redSum - blueSum)
}

function shuffleArray(arr) {
  const a = [...arr]
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]]
  }
  return a
}

/**
 * 选取黑原子破坏目标（攻击/红效果通用）。
 * 规则：要破坏 x 个黑原子时：
 * - 若无黄高亮黑原子：从该格所有黑原子中无放回随机抽 x 个作为目标；
 * - 若有 y 个黄高亮黑原子且 x>=y：先确定这 y 个为目标，再从其余黑原子中随机抽 x-y 个；
 * - 若有 y 个黄高亮黑原子且 x<y：从这 y 个中随机抽 x 个作为目标。
 * 选中的目标若被蓝保护则不实际破坏。
 * 返回 { results: [{ ptKey, destroyed }], destroyedAtoms: [{ defender, cellIndex, r, c, color }] }
 */
export function selectAndDestroyBlackTargets(state, defender, cellIndex, defCell, x, actuallyDestroy = true) {
  const allBlacks = [...defCell.blackPoints()]
  const yellowPriority = allBlacks.filter((k) => state.yellowPriorityPoints?.[defender]?.has(`${cellIndex}:${k}`) ?? false)
  const otherBlacks = allBlacks.filter((k) => !yellowPriority.includes(k))
  const y = yellowPriority.length

  let targets = []
  if (y === 0) {
    targets = shuffleArray(allBlacks).slice(0, Math.min(x, allBlacks.length))
  } else if (x >= y) {
    targets = [...shuffleArray(yellowPriority)]
    const needMore = x - y
    if (needMore > 0 && otherBlacks.length > 0) {
      targets.push(...shuffleArray(otherBlacks).slice(0, Math.min(needMore, otherBlacks.length)))
    }
  } else {
    targets = shuffleArray(yellowPriority).slice(0, Math.min(x, yellowPriority.length))
  }

  const results = []
  const destroyedAtoms = []
  for (const ptKey of targets) {
    const protected_ = state.blueProtectedPoints?.[defender]?.has(`${cellIndex}:${ptKey}`) ?? false
    if (actuallyDestroy && !protected_) {
      const [r, c] = ptKey.split(',').map(Number)
      defCell.remove(r, c)
      results.push({ ptKey, destroyed: true })
      destroyedAtoms.push({ defender, cellIndex, r, c, color: 'black' })
    } else {
      results.push({ ptKey, destroyed: !protected_ })
    }
  }
  return { results, destroyedAtoms }
}

export function destroyOneBlackAndGetComponents(attackerCell, defenderCell, blackToDestroy) {
  const [r, c] = Array.isArray(blackToDestroy) ? blackToDestroy : blackToDestroy.split(',').map(Number)
  if (defenderCell.get(r, c) !== ATOM_BLACK) return [0, []]
  defenderCell.remove(r, c)
  const components = defenderCell.connectedComponents()
  return [1, components]
}

export function removeComponentsExcept(cell, toKeep) {
  const components = cell.connectedComponents()
  for (const comp of components) {
    let overlap = false
    for (const k of comp) if (toKeep.has(k)) { overlap = true; break }
    if (overlap) continue
    for (const k of comp) {
      const [r, c] = k.split(',').map(Number)
      cell.remove(r, c)
    }
  }
}

export function removeComponent(cell, component) {
  for (const k of component) {
    const [r, c] = k.split(',').map(Number)
    cell.remove(r, c)
  }
}

export function removeBlackAtomsExcept(cell, toKeep) {
  const blacks = cell.blackPoints()
  for (const k of blacks) {
    if (!toKeep.has(k)) {
      const [r, c] = k.split(',').map(Number)
      cell.remove(r, c)
    }
  }
}

export function clearCellIfNoBlack(cell) {
  if (!cell.hasBlack() && !cell.isEmpty()) {
    const toRemove = [...cell.allAtoms()].map(([pt]) => pt)
    for (const [r, c] of toRemove) cell.remove(r, c)
  }
}

/**
 * 破坏后连通规则（进攻与红效果均按此流程）：
 * (1) 检查该格剩余原子是否连通，不连通则返回 { type: 'all' } 由被破坏方选择保留一个连通子集；
 * (2) 仅看黑原子是否连通，若存在多个黑原子连通子集则返回 { type: 'black' } 由被破坏方选择保留一个；
 * (3) 再自动清除所有「不包含黑原子」的连通子集。
 * 返回 null 表示无需选择或已执行完 (3)；否则返回 { type, components }，components 为可序列化的 [string[]]。
 */
export function getConnectivityChoice(cell) {
  if (cell.isEmpty()) return null
  // (1) 全原子连通
  const comps = cell.connectedComponents()
  if (comps.length > 1) {
    return { type: 'all', components: comps.map((s) => [...s]) }
  }
  // (2) 黑原子连通
  const blackComps = cell.blackConnectedComponents()
  if (blackComps.length > 1) {
    return { type: 'black', components: blackComps.map((s) => [...s]) }
  }
  // (3) 自动清除不含黑的连通子集
  removeComponentsWithoutBlackAndReturnRest(cell)
  return null
}

/**
 * 被破坏方选择后应用：保留 componentToKeepKeys（string[]），移除其余。
 * type 'all'：保留一个全原子连通子集；type 'black'：保留一个黑原子连通子集并执行步骤(3)。
 */
export function applyConnectivityChoice(cell, type, componentToKeepKeys) {
  const toKeep = new Set(componentToKeepKeys)
  if (type === 'all') {
    removeComponentsExcept(cell, toKeep)
  } else if (type === 'black') {
    removeBlackAtomsExcept(cell, toKeep)
    removeComponentsWithoutBlackAndReturnRest(cell) // (3) 清除不含黑的连通子集
  }
}

/**
 * 破坏后连通规则（随机版，用于无需玩家选择时）。
 */
export function applyPostDestructionConnectivity(cell) {
  if (cell.isEmpty()) return
  let comps = cell.connectedComponents()
  if (comps.length > 1) {
    const toKeep = comps[Math.floor(Math.random() * comps.length)]
    removeComponentsExcept(cell, toKeep)
  }
  if (cell.isEmpty()) return
  let blackComps = cell.blackConnectedComponents()
  if (blackComps.length > 1) {
    const toKeep = blackComps[Math.floor(Math.random() * blackComps.length)]
    removeBlackAtomsExcept(cell, toKeep)
  }
  removeComponentsWithoutBlackAndReturnRest(cell)
}

export function removeComponentsWithoutBlackAndReturnRest(cell) {
  if (cell.isEmpty()) return []
  const components = cell.connectedComponents()
  const blacks = cell.blackPoints()
  const remaining = []
  for (const comp of components) {
    let hasBlack = false
    for (const k of comp) if (blacks.has(k)) { hasBlack = true; break }
    if (hasBlack) remaining.push(comp)
    else {
      for (const k of comp) {
        const [r, c] = k.split(',').map(Number)
        cell.remove(r, c)
      }
    }
  }
  return remaining
}

export function resolveDirectAttack(cell) {
  return Math.floor(attackPower(cell))
}

export function applyEffectBlue(state, player, cellIndex, r, c) {
  const cells = state.cells[player]
  if (cellIndex < 0 || cellIndex >= cells.length) return false
  const cell = cells[cellIndex]
  if (cell.get(r, c) !== ATOM_BLUE) return false
  const protected_ = cell.effectiveBlackNeighborsOf(r, c)
  const purpleNeighbors = cell.purpleNeighborPositions(r, c)
  cell.remove(r, c)
  for (const [nr, nc] of purpleNeighbors) {
    cell.remove(nr, nc)
  }
  for (const k of protected_) {
    state.blueProtectedPoints[player].add(`${cellIndex}:${k}`)
  }
  state.blueProtectionUntilTurn[player] = state.turnNumber + 2
  return true
}

/**
 * 红原子点击效果：先选对方格子，再在该格内随机破坏 y 个黑原子（y=红原子相邻黑数）。
 * - 从 UI 调用时必须传入 targetDefCellIndex（用户选择的对方格索引）。
 * - 仅从该格内收集可破坏的黑原子，随机取 min(y, 该格内数量) 个破坏。
 */
export function applyEffectRedRandom(state, attacker, cellIndex, r, c, targetDefCellIndex = null) {
  const cells = state.cells[attacker]
  if (cellIndex < 0 || cellIndex >= cells.length) return false
  const cell = cells[cellIndex]
  if (cell.get(r, c) !== ATOM_RED) return false
  const y = cell.effectiveBlackNeighborCount(r, c)
  if (y <= 0) return false
  const defender = 1 - attacker
  const defCells = state.cells[defender]
  // 仅从选中的对方格（或未指定时全部格）收集候选
  const cellIndices =
    targetDefCellIndex != null && targetDefCellIndex >= 0 && targetDefCellIndex < defCells.length
      ? [targetDefCellIndex]
      : targetDefCellIndex == null
        ? Array.from({ length: defCells.length }, (_, i) => i)
        : []
  cell.remove(r, c)
  const affectedCells = new Set()
  const destroyedAtoms = []
  for (const ci of cellIndices) {
    const defCell = defCells[ci]
    const numBefore = defCell.allAtoms().length
    if (numBefore === 0) continue
    const numBlacks = defCell.blackPoints().size
    if (numBlacks === 0) continue
    const k = Math.min(y, numBlacks)
    const { destroyedAtoms: da } = selectAndDestroyBlackTargets(state, defender, ci, defCell, k, true)
    destroyedAtoms.push(...da)
    if (defCell.allAtoms().length < numBefore) affectedCells.add(ci)
  }
  // 红效果破坏结束后立即检查是否产生多个连通子集，若有则需被破坏方选择
  for (const ci of affectedCells) {
    const choice = getConnectivityChoice(defCells[ci])
    if (choice) {
      return { ok: true, connectivityChoice: { defender, cellIndex: ci, ...choice }, destroyedAtoms }
    }
  }
  return { ok: true, destroyedAtoms }
}

export function applyEffectYellow(state, player, cellIndex, r, c) {
  const cells = state.cells[player]
  if (cellIndex < 0 || cellIndex >= cells.length) return false
  const cell = cells[cellIndex]
  if (cell.get(r, c) !== ATOM_YELLOW) return false
  const priority_ = cell.effectiveBlackNeighborsOf(r, c)
  const purpleNeighbors = cell.purpleNeighborPositions(r, c)
  cell.remove(r, c)
  for (const [nr, nc] of purpleNeighbors) {
    cell.remove(nr, nc)
  }
  for (const k of priority_) {
    state.yellowPriorityPoints[player].add(`${cellIndex}:${k}`)
  }
  state.yellowPriorityUntilTurn[player] = state.turnNumber + 2
  return true
}

export function applyEffectGreen(state, player, cellIndex, r, c) {
  const cells = state.cells[player]
  if (cellIndex < 0 || cellIndex >= cells.length) return false
  const cell = cells[cellIndex]
  if (cell.get(r, c) !== ATOM_GREEN) return false
  const extendToEmpty = cell.hasPurpleNeighbor(r, c)
  const oneHopBlack = extendToEmpty ? cell.effectiveBlackNeighborsOf(r, c) : new Set()
  const purpleNeighbors = cell.purpleNeighborPositions(r, c)
  cell.remove(r, c)
  cell.place(r, c, ATOM_BLACK)
  if (extendToEmpty) {
    for (const key of oneHopBlack) {
      const [r1, c1] = key.split(',').map(Number)
      for (const [nr, nc] of cell.grid.neighborsOf(r1, c1)) {
        if (cell.get(nr, nc) == null && cell.grid.inBounds(nr, nc)) {
          cell.place(nr, nc, ATOM_BLACK)
        }
      }
    }
  }
  for (const [nr, nc] of purpleNeighbors) {
    cell.remove(nr, nc)
  }
  return true
}

export function applyGreenEndOfTurn(state, player) {
  let total = 0
  for (const cell of state.cells[player]) {
    for (const [[r, c], color] of cell.allAtoms()) {
      if (color === ATOM_GREEN) total += cell.effectiveBlackNeighborCount(r, c)
    }
  }
  if (total > 0) {
    state.pools[player][ATOM_BLACK] = (state.pools[player][ATOM_BLACK] ?? 0) + total
  }
  return total
}
